hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/txns/crypto/HollowAccountFinalizationLogic.java

```diff
@@ -0,0 +1,184 @@
+            final SigImpactHistorian sigImpactHistorian,
+            final RecordsHistorian recordsHistorian,
+            final ExpandHandleSpanMapAccessor spanMapAccessor,
+            final AliasManager aliasManager,
+            final GlobalDynamicProperties properties) {
+        this.accountsSupplier = accountsSupplier;
+        this.creator = creator;
+        this.syntheticTxnFactory = syntheticTxnFactory;
+        this.sigImpactHistorian = sigImpactHistorian;
+        this.recordsHistorian = recordsHistorian;
+        this.spanMapAccessor = spanMapAccessor;
+        this.aliasManager = aliasManager;
+        this.properties = properties;
+        this.txnCtx = txnCtx;
+    }
+
+    // add javadocs
+    public ResponseCodeEnum performFor(final Map<byte[], JKey> activeEvmAddressesToECDSAKeys) {
+        final var pendingFinalizations = new HashMap<HederaAccount, JKey>();
+
+        final var status =
+                populatePendingFinalizationsFrom(
+                        activeEvmAddressesToECDSAKeys, pendingFinalizations);
+        if (status != OK) {
+            return status;
+        }
+
+        finalizeAccounts(pendingFinalizations);
+        return OK;
+    }
+
+    private ResponseCodeEnum populatePendingFinalizationsFrom(
+            final Map<byte[], JKey> activeEvmAddressesToECDSAKeys,
+            final Map<HederaAccount, JKey> pendingFinalizations) {
+        if (activeEvmAddressesToECDSAKeys != null) {
+            for (final var entry : activeEvmAddressesToECDSAKeys.entrySet()) {
+                final var status =
+                        addIfNeeded(entry.getKey(), entry.getValue(), pendingFinalizations);
+                if (status != OK) {
+                    return status;
+                }
+            }
+        }
+        // check if wrapped Ethereum txn sender is present and needs finalizing
+        final var accessor = txnCtx.accessor();
+        final var ethTxExpansion = spanMapAccessor.getEthTxExpansion(accessor);
+        if (properties.isLazyCreationEnabled()
+                && ethTxExpansion != null
+                && ethTxExpansion.result().equals(OK)) {
+            final var ethTxSigs = spanMapAccessor.getEthTxSigsMeta(accessor);
+            return addIfNeeded(
+                    ethTxSigs.address(),
+                    new JECDSASecp256k1Key(ethTxSigs.publicKey()),
+                    pendingFinalizations);
+        }
+        return OK;
+    }
+
+    private ResponseCodeEnum addIfNeeded(
+            final byte[] address,
+            final JKey key,
+            final Map<HederaAccount, JKey> pendingFinalizations) {
+        final var accountNum = aliasManager.lookupIdBy(wrapUnsafely(address));
+        if (accountNum != EntityNum.MISSING_NUM) {
+            // DISCUSS --- getForModify directly, or get() first, and then getForModify() if needed
+            final var hederaAccount = accountsSupplier.get().getForModify(accountNum);
```

Comment:
I imagine **most of the cases** where we have a valid `ECDSA` key signature there won't be any hollow finalizations, so the  `getForModify` here won't be needed. Will there be a performance hit of always calling it, even if not needed, I'm not familiar enough with the state implementation @Neeharika-Sompalli @tinker-michaelj